x/ratelimit/rate_limit.go

```diff
@@ -0,0 +1,155 @@
+//              Packet Denom:      transfer/channel-X/ustrd
+//               -> Remove Prefix: ustrd
+//               -> Leave as is:   ustrd
+//
+//			Ex2: juno was sent to Stride, then to Osmosis, then back to Stride
+//              Packet Denom:      transfer/channel-X/transfer/channel-Z/ujuno
+//               -> Remove Prefix: transfer/channel-Z/ujuno
+//               -> Hash:          ibc/...
+func ParseDenomFromRecvPacket(packet channeltypes.Packet, packetData transfertypes.FungibleTokenPacketData) (denom string) {
+	// To determine the denom, first check whether Stride is acting as source
+	if transfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), packetData.Denom) {
+		// Remove the source prefix (e.g. transfer/channel-X/transfer/channel-Z/ujuno -> transfer/channel-Z/ujuno)
+		sourcePrefix := transfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+		unprefixedDenom := packetData.Denom[len(sourcePrefix):]
+
+		// Native assets will have an empty trace path and can be returned as is
+		denomTrace := transfertypes.ParseDenomTrace(unprefixedDenom)
+		if denomTrace.Path == "" {
+			denom = unprefixedDenom
+		} else {
+			// Non-native assets should be hashed
+			denom = denomTrace.IBCDenom()
+		}
+	} else {
+		// Prefix the destination channel - this will contain the trailing slash (e.g. transfer/channel-X/)
+		destinationPrefix := transfertypes.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
+		prefixedDenom := destinationPrefix + packetData.Denom
+
+		// Hash the denom trace
+		denomTrace := transfertypes.ParseDenomTrace(prefixedDenom)
+		denom = denomTrace.IBCDenom()
+	}
+
+	return denom
+}
+
+// Middleware implementation for SendPacket with rate limiting
+func SendRateLimitedPacket(ctx sdk.Context, keeper ratelimitkeeper.Keeper, packet exported.PacketI) error {
+	// For a send packet, the channel on stride is the "Source" channel
+	//  This is because the Source and Desination are defined from the perspective of a packet recipient
+	//    i.e., when this packet lands on a the host chain, the "Source" will show the Stride Channel
+	channelId := packet.GetSourceChannel()
+
+	// Parse the packet data
+	var packetData transfertypes.FungibleTokenPacketData
+	if err := json.Unmarshal(packet.GetData(), &packetData); err != nil {
+		return err
+	}
+
+	amount, ok := sdk.NewIntFromString(packetData.Amount)
+	if !ok {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Unable to cast packet amount to sdk.Int")
+	}
+
+	denom := ParseDenomFromSendPacket(packetData)
+
+	err := keeper.CheckRateLimitAndUpdateFlow(ctx, types.PACKET_SEND, denom, channelId, amount)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Middleware implementation for RecvPacket with rate limiting
+func ReceiveRateLimitedPacket(ctx sdk.Context, keeper ratelimitkeeper.Keeper, packet channeltypes.Packet) error {
+	// For a receive packet, the channel on stride is the "Destination" channel
```

Comment:
Can you add more detail to this sentence? I think the implication here is that the Stride channel is used to key `RateLimit` objects

Reply:
yeah that's exactly it.

Added some clarifying comments